### PR TITLE
Experimental: add new WiFi (pseudo) modes: WIFI_SHUTDOWN & WIFI_RESUME

### PR DESCRIPTION
**Pull request by d-a-v**
*edit*: This is an experimental addition to the API

Adds new WiFi (pseudo) modes: WIFI_SHUTDOWN & WIFI_RESUME,
also:
- an example
- restore WiFi.onWiFiModeChange()
- add crc32() in core
- cancel #5703 because of https://github.com/esp8266/Arduino/issues/5667#issuecomment-512180419 (related to all the issues referenced below)

references:
#6346 
#6266 
#6057 #5527

This is an attempt to put right into the core all the necessary code to wifi-sleep and deep-sleep so there is only one place to debug for everybody.
I am personally not able to reproduce all the WDT issues so even after hypothetical merge, we will need to bring fixes found by all the community, like  https://github.com/esp8266/Arduino/issues/6266#issuecomment-515803290 and https://github.com/esp8266/Arduino/issues/6172#issuecomment-500457407 (only citing them).

This PR does not change legacy behavior when using the regular API (well.. not supposed to :]).

### CODE REVIEW

File: libraries/ESP8266WiFi/src/ESP8266WiFiGeneric.cpp [hunk 1]
```diff
@@ -386,22 +394,36 @@ bool ESP8266WiFiGenericClass::getPersistent(){
...
+
+    if (wifi_fpm_get_sleep_type() != NONE_SLEEP_T) {
+        // wifi may have been put asleep by ESP8266WiFiGenericClass::preinitWiFiOff
+        wifi_fpm_do_wakeup();
+        wifi_fpm_close();
+    }
+
     if(_persistent){
         if(wifi_get_opmode() == (uint8) m && wifi_get_opmode_default() == (uint8) m){
+            saveState(state, m, _persistent);
```
Comment (TD-er) [inline libraries/ESP8266WiFi/src/ESP8266WiFiGeneric.cpp]: I don't get it.
If the mode is already the same as the one you want to set, then save the state?

File: libraries/ESP8266WiFi/src/include/WiFiState.h [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    struct
+    {
+        bool persistent;
+        WiFiMode_t mode;
+        uint8_t channel;
+        station_config fwconfig;
+        ip_info ip;
+        ip_addr_t dns[2];
+        ip_addr_t ntp[2];
+    } state;
```
Comment (TD-er) [inline libraries/ESP8266WiFi/src/include/WiFiState.h]: Ordering members from big to small will often lead to smaller struct sizes.
Also the members are not initialized in the constructor.
For example channel is only set if there was a successful connect and/or the active mode was WIFI_STA when shutdown and fwconfig is only set when in STA mode at shutdown.
But these are being used at wifi.begin.

I have not looked in much detail to see if the settings are somewhere read from flash, but I can imagine that would result in very hard to reproduce results.

File: libraries/ESP8266WiFi/src/ESP8266WiFiGeneric.cpp (hunk 1 above)
Comment (d-a-v) [reply libraries/ESP8266WiFi/src/ESP8266WiFiGeneric.cpp]: This is right.
saveState should be used only with shutdown (write all structure members) or resume mode (real). Fixed.

File: libraries/ESP8266WiFi/src/include/WiFiState.h (hunk 2 above)
Comment (d-a-v) [reply libraries/ESP8266WiFi/src/include/WiFiState.h]: Members are initialized in shutdown only (now). The whole structure is valid only when crc is matching, and crc is calculated when all arguments are set.

File: libraries/ESP8266WiFi/src/ESP8266WiFiType.h [hunk 3]
```diff
@@ -58,9 +59,10 @@ typedef enum WiFiEvent
     WIFI_EVENT_SOFTAPMODE_STACONNECTED,
     WIFI_EVENT_SOFTAPMODE_STADISCONNECTED,
     WIFI_EVENT_SOFTAPMODE_PROBEREQRECVED,
+    WIFI_EVENT_MODE_CHANGE,
```
Comment (devyte) [inline libraries/ESP8266WiFi/src/ESP8266WiFiType.h]: Maybe swap these to keep the softapmode together

Comment (d-a-v) [reply libraries/ESP8266WiFi/src/ESP8266WiFiType.h]: The order needs to be the same as [SDK defines](https://github.com/esp8266/Arduino/blob/0399bb1fee4b95bfb8f9007f561e4f7060eedcdb/tools/sdk/include/user_interface.h#L466)'s

File: cores/esp8266/crc32.cpp [hunk 4]
```diff
@@ -0,0 +1,23 @@
+
+#include "coredecls.h"
+
+// taken from esp8266/examples/RTCUserMemory/RTCUserMemory.ino
+uint32_t crc32 (const void* data, size_t length)
```
[the remark this answers is not in the document]
Comment (d-a-v) [reply cores/esp8266/crc32.cpp]: There is no difference. But eboot is a standalone code and only uses SDK.
Sharing common source code with eboot should be addressed from another issue.

File: libraries/ESP8266WiFi/src/ESP8266WiFiGeneric.h [hunk 5]
```diff
@@ -88,6 +90,11 @@ class ESP8266WiFiGenericClass {
         bool forceSleepBegin(uint32 sleepUs = 0);
         bool forceSleepWake();
 
+        bool shutdown (uint32 sleepUs = 0, WiFiState* stateSave = nullptr);
```
[the remark this answers is not in the document]
Comment (d-a-v) [reply libraries/ESP8266WiFi/src/ESP8266WiFiGeneric.h]: Commented and moved to protected

**Pull request merged**